src/lib/bootstrap_controller/bootstrap_controller.ml

```diff
@@ -0,0 +1,202 @@
+      done_syncing_root t
+      || (not @@ worth_getting_root t candidate time_received)
+    then Deferred.unit
+    else
+      match%map
+        Network.get_ancestry t.network (input.descendant, input.generations)
+      with
+      | Error e ->
+          Logger.error t.logger
+            !"Could not get the proof of ancestors from the \
+              network:%{sexp:Error.t}"
+            e
+      | Ok (ancestor, proof) -> (
+          let ancestor_length =
+            Protocol_state.(Consensus_state.length (consensus_state ancestor))
+          in
+          match
+            Ancestor.Prover.verify_and_add t.ancestor_prover input
+              (Protocol_state.hash ancestor)
+              proof ~ancestor_length
+          with
+          | Ok () ->
+              t.best_with_root <- {state= candidate; root= ancestor} ;
+              let candidate_body_hash =
+                Protocol_state.Body.hash (Protocol_state.body candidate)
+              in
+              let candidate_hash = Protocol_state.hash candidate in
+              Ancestor.Prover.add t.ancestor_prover ~hash:candidate_hash
+                ~prev_hash:previous_state_hash
+                ~length:
+                  Protocol_state.(
+                    Consensus_state.length (consensus_state candidate))
+                ~body_hash:candidate_body_hash ;
+              let ledger_hash =
+                Consensus.Mechanism.(
+                  Protocol_state.blockchain_state ancestor
+                  |> Blockchain_state.ledger_hash
+                  |> Frozen_ledger_hash.to_ledger_hash)
+              in
+              Syncable_ledger.new_goal t.syncable_ledger ledger_hash |> ignore
+          | Error e -> received_bad_proof t e )
+
+  (* TODO: We need to do catchup jobs for all remaining transitions in the cache. 
+           This will be hooked into `run` when we do this *)
+  let _expand_root ~frontier root_hash cache =
+    let rec dfs state_hash =
+      Option.iter (Hashtbl.find_and_remove cache state_hash)
+        ~f:(fun children ->
+          List.iter children ~f:(fun transition ->
+              Transition_frontier.add_transition_exn frontier transition
+              |> ignore ;
+              dfs (With_hash.hash transition) ) )
+    in
+    dfs root_hash
+
+  let sync_ledger t ~transition_graph ~transition_reader =
+    Reader.iter transition_reader
+      ~f:(fun (`Transition incoming_transition, `Time_received time_received)
+         ->
+        let transition = Envelope.Incoming.data incoming_transition in
+        let protocol_state = External_transition.protocol_state transition in
+        let previous_state_hash =
+          External_transition.Protocol_state.previous_state_hash protocol_state
+        in
+        Transition_cache.add transition_graph ~parent:previous_state_hash
+          transition ;
+        if worth_getting_root t protocol_state time_received then
+          on_transition t (transition, time_received) |> don't_wait_for ;
```

Comment:
this `don't_wait_for` can spin up a new green thread, this is where you need a pipe write instead

Reply:
but actually sync with @nholland94 because he's building a similar abstraction to what you'll want here

Reply:
I'm adding a `Capped_supervisor` in my current branch that will add an easy interface for this functionality. I should have that up soon.